OP-1020 | DELETE last movement in Pharmaceutical Stock (#238)
informatici/openhospital-doc: doc_user/UserManual.adoc

Reviewer: The NOTE at line 500 is still there, please remove it.

* *C[.underline]##h##arge*: record a new charge movement.
* *[.underline]##D##ischarge*: record a new discharge movement.

> * *[.underline]##D##elete*: delete the last movement.

Reviewer: The underlined character should be the one used as _Mnemonic_ in the code and cannot be shared with other buttons in the same window.

According with the current code in develop now, the Delete button has been mapped with common `angal.common.delete.btn.key=D`, but in this windows we might map differently because we already have `angal.medicalstock.discharge.btn.key=D`, so we could define new `angal.medicalstock.discharge.btn=Delete` and `angal.medicalstock.delete.btn.key=L` or having `angal.medicalstock.discharge.btn.key=R`.

Since we did the same for Charge and Close button (keeping the "C" for common operation "Close") I would assume the best choice is to keep "D" for common operation of "Delete" and having Charge=H and Discharge=R.

In any case this doc should be updated accordingly.

Author: please should I apply those changes now so that after you will create the task to update it on gui ?

Reviewer: we are still in the matter of OP-1020, just create a new branch from the latest develop and open a new PR for it

Author: and this branch must contain what I don't understand please.

Reviewer: > I would assume the best choice is to keep "D" for common operation of "Delete" and having Charge=H and Discharge=R.

this

Author: Okay the task is to just change the  key of charge and discharge button

movement, several discharging movements may be generated. Ask the Administrator or check the _Administrator’s Guide_
for more information.

[#delete-last-stock-movement]
===== 4.2.2.3 Delete Last Stock Movement (*[.underline]##D##elete*)

> To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. if it is not the last movement you will have and *_Warning_* window:

Reviewer 2: ```suggestion
To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. If it is not the last movement you will see a *_Warning_* window:
```

> image:image250.png[Warning]

Reviewer 2: This should be an ERROR and not just a warning; see dialog title.

Author: the title is Warning but should I rename it to error ?

Reviewer 2: I think so as a warning to me means something to know about but we will continue anyway.   Doesn't this really mean you can't do it?  So it should be IMHO an error.

Author: okay you right but we already merge the gui so can I create the task to update the window and use error dialog instead of warning dialog?

Reviewer: It is actually just warning even before the operation takes place, so I think it is different from an ERROR which comes during an operation, but I think the point here is consistency with the rest of the doc.

According with the PR merged in GUI it is a Warning, maybe we can have it consistent with the GUI as for now and change it later on?

Reviewer: Anyway, `image250` it is showing only if no movement is selected.

The correct dialog shown when the user select a "not-last movement", it's still a warning but like this one:

![image](https://github.com/informatici/openhospital-doc/assets/2938553/b8ddf39d-83b5-447f-b903-04d301d67eef)

Reviewer: I've checked, and according with the rest of the code and documentation, we used warning only few times and for not-blocking operations, we used error for all other cases instead.

So @ArnaudFonzam you can include this change in the new PR on GUI, thanks!

Author: Okay

Reviewer: and update this doc accordingly, thanks!

Author: okay I'm working on it

> But, if the row selected is the last movement you will have the *_Question_* window:

Reviewer 2: ```suggestion
But if the row selected is the last movement the *_Question_* window is presented:
```

image:image248.png[Question]

> Click *[.underline]##Y##es* if you realy want to delete the last movement or click *[.underline]##N##o* if no. Before click the *[.underline]##Y##es* button to be sure that the product present in the last movement is not used in the *_Pharmaceuticals Stock [.underline]##W##ard_*

Reviewer 2: ```suggestion
Click *[.underline]##Y##es* to confirm the deletion of the last movement or click *[.underline]##N##o* to not  delete the movement. Before clicking the *[.underline]##Y##es* button check that the product present in the last movement is not used in the *_Pharmaceuticals Stock [.underline]##W##ard_*
```

Reviewer 2: The documentation tries to avoid at all cost the use of *you*

> if yes you will have the *_Error_* window.

Reviewer 2: ```suggestion
if it is present then the *_Error_* window is shown.
```

image:image249.png[Error]

> If no, the last stock movement will be deleted from the table and you will have the *_Success_* window:

Reviewer 2: ```suggestion
If it is not present then the last stock movement is deleted from the table and the *_Success_* window is shown:
```

image:image251.png[success]

> NOTE: This button can be deactivated if you don't want to allow movements to be deleted. Ask the Administrator or check the Administrator's Guide for more information.

Reviewer 2: ```suggestion
NOTE: This button can be deactivated to not allow movements to be deleted. Ask the Administrator or check the Administrator's Guide for more information.
```

movement, several discharging movements may be generated. Ask the Administrator or check the _Administrator’s Guide_
for more information.

> <<<

Reviewer: revert this deletion, this symbol is needed for the PDF pagination

===== 4.2.2.3 Delete Last Stock Movement (*[.underline]##D##elete*)
To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. If it is not the last movement you will see a *_Error_* window:

image:image250.png[Error]

if you don't choosed any row and:

image:image252.png[Error]

if you choosed more than one row.

> But if the row selected is the last movement the *_Question_* window is presented:

Reviewer: Let's simplify this part. First of all we have to remove the NOTE in line 500. Then, as already suggested, we remove all "you" occurences and make it all impersonal. Then I would rephrase it as follow, and avoid to explain all possible errors. 

Please upload the _missing image_.

```suggestion
In Pharmaceutical Stock it is not possible to change the past and all operations need to be done in consecutive order. If a mistake concern a movement in the past (for example, the wrong quantity of a pharmaceutical is inserted in charge), then use a new opposite movement to correct the mistake (in the example, use a discharge movement to undo the unwanted change).

Nevertheless, the admin can enable/disable the *[.underline]##D##elete* button that allows the enabled user do delete only the last movement, and helps quickly undoing mistakes.

To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. Only one row is allowed to delete at a time.

If it is not the last movement the following  *_Error_* message is shown:

image:<missing image>[Not last movement error]

But if the row selected is the last movement this *_Question_* message is presented:
```

image:image248.png[Question]

> Click *[.underline]##Y##es* if you want to confirm the deletion of the last movement or click *[.underline]##N##o* if you do not want to delete the movement. Before clicking the *[.underline]##Y##es* button check that the product present in the last movement is not used in the *_Pharmaceuticals Stock [.underline]##W##ard_*
> if it is then the *_Error_* window is shown.
> 
> image:image249.png[Error]
> 
> If it is not present then the last stock movement will be deleted from the table and the *_Success_* window is shown:
> 
> image:image251.png[success]
> 
> NOTE: This button can be deactivated to not allow movements to be deleted. Ask the Administrator or check the Administrator's Guide for more information.

Reviewer: ```suggestion
Click *[.underline]##Y##es* to confirm the deletion of the last movement, otherwise click *[.underline]##N##o* to cancel the operation. After clicking the *[.underline]##Y##es* button, the system checks that the product involved the selected movement has not been used itself in the *_Pharmaceuticals Stock [.underline]##W##ard_*.

If the product has been already used in a Ward, then the *_Error_* message is shown.

image:image249.png[Error]

Otherwise the selected movement is deleted and the *_Success_* window is shown:

image:image251.png[success]

NOTE: This button can be enabled/disabled to allow/not allow movements deletion. Ask the Administrator or check the Administrator's Guide for more information.
```

<<<

[#delete-last-stock-movement]
===== 4.2.2.3 Delete Last Stock Movement (*[.underline]##D##elete*)

> In Pharmaceutical Stock it is not possible to change the past and all operations need to be done in consecutive order. If a mistake concern a movement in the past (for example, the wrong quantity of a pharmaceutical is inserted in charge), then use a new opposite movement to correct the mistake (in the example, use a discharge movement to undo the unwanted change).

Reviewer: ```suggestion
In Pharmaceutical Stock it is not possible to change the past and all operations need to be done in consecutive order. If a mistake concerns a movement in the past (for example, the wrong quantity of a pharmaceutical is inserted in charge), then use a new opposite movement to correct the mistake (in the example, use a discharge movement to undo the unwanted change).
```

Nevertheless, the admin can enable/disable the *[.underline]##D##elete* button that allows the enabled user do delete only the last movement, and helps quickly undoing mistakes.

To delete the last movement, select first the row that you want to delete and press the *[.underline]##D##elete* button in the *_Stock Movement Browser_* window. Only one row is allowed to delete at a time.

If it is not the last movement the following  *_Error_* message is shown:

> image:image250.png[Not last movement error]

Reviewer: Please this image is still the wrong one

But if the row selected is the last movement this *_Question_* message is presented:

image:image248.png[Question]

> Click *[.underline]##Y##es* to confirm the deletion of the last movement, otherwise click *[.underline]##N##o* to cancel the operation. After clicking the *[.underline]##Y##es* button, the system checks that the product involved the selected movement has not been used itself in the *_Pharmaceuticals Stock [.underline]##W##ard_*.

Reviewer: ```suggestion
Click *[.underline]##Y##es* to confirm the deletion of the last movement, otherwise click *[.underline]##N##o* to cancel the operation. After clicking the *[.underline]##Y##es* button, the system checks if the product involved in the selected movement has not been used itself in the *_Pharmaceuticals Stock [.underline]##W##ard_* (see <<pharmaceutical-stock-ward,Pharmaceuticals Stock Ward>> in this document).
```
